File: app/src/main/java/net/typeblog/socks/AppSelector.java
package net.typeblog.socks;

import android.app.Activity;
import android.app.AlertDialog;
import android.content.Context;
import android.content.pm.ApplicationInfo;
import android.content.pm.PackageManager;
import android.graphics.drawable.Drawable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.CheckBox;
import android.widget.EditText;
import android.widget.Filter;
import android.widget.Filterable;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.ListView;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

public class AppSelector {

    public interface OnAppSelectedListener {
        void onAppSelected(String appList);
    }

    public static void show(Activity activity, String currentApps, final OnAppSelectedListener listener) {
        final PackageManager pm = activity.getPackageManager();
        List<ApplicationInfo> packages = pm.getInstalledApplications(PackageManager.GET_META_DATA);
        final List<AppInfo> appList = new ArrayList<>();
        
        Set<String> selectedPackages = new HashSet<>();
        if (currentApps != null && !currentApps.isEmpty()) {
            for (String p : currentApps.split("\n")) {
                selectedPackages.add(p.trim());
            }
        }

        for (ApplicationInfo packageInfo : packages) {
            // Filter out system apps that are not important or don't have a launcher? 
            // Better to show all but sort them.
            AppInfo info = new AppInfo();
            info.name = packageInfo.loadLabel(pm).toString();
            info.packageName = packageInfo.packageName;
            info.icon = packageInfo.loadIcon(pm);
            info.selected = selectedPackages.contains(info.packageName);
            appList.add(info);
        }

        Collections.sort(appList, new Comparator<AppInfo>() {
            @Override
            public int compare(AppInfo o1, AppInfo o2) {
                return o1.name.compareToIgnoreCase(o2.name);
            }
        });

        LinearLayout layout = new LinearLayout(activity);
        layout.setOrientation(LinearLayout.VERTICAL);

        EditText searchBox = new EditText(activity);
        searchBox.setHint(R.string.search_apps);
        searchBox.setSingleLine(true);
        layout.addView(searchBox);

        ListView listView = new ListView(activity);
        final AppAdapter adapter = new AppAdapter(activity, appList);
        listView.setAdapter(adapter);
        
        // Add ListView with weight 1 so it takes remaining space
        LinearLayout.LayoutParams listParams = new LinearLayout.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT, 0, 1.0f);
        layout.addView(listView, listParams);

        searchBox.addTextChangedListener(new android.text.TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {}
            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                adapter.getFilter().filter(s);
            }
            @Override
            public void afterTextChanged(android.text.Editable s) {}
        });

        new AlertDialog.Builder(activity)
                .setTitle(R.string.adv_app_selector)
                .setView(layout)
                .setPositiveButton(android.R.string.ok, (dialog, which) -> {
                    // Use simple loop for compatibility
                    StringBuilder result = new StringBuilder();
                    for (AppInfo info : appList) {
                        if (info.selected) {
                            result.append(info.packageName).append("\n");
                        }
                    }
                    listener.onAppSelected(result.toString().trim());
                })
                .setNegativeButton(android.R.string.cancel, null)
                .show();
    }

    private static class AppInfo {
        String name;
        String packageName;
        Drawable icon;
        boolean selected;
    }

    private static class AppAdapter extends BaseAdapter implements Filterable {
        private Context context;
        private List<AppInfo> appsOriginal;
        private List<AppInfo> appsFiltered;
        private AppFilter filter;

        AppAdapter(Context context, List<AppInfo> apps) {
            this.context = context;
            this.appsOriginal = apps;
            this.appsFiltered = apps;
        }

        @Override
        public int getCount() {
            return appsFiltered.size();
        }

        @Override
        public Object getItem(int position) {
            return appsFiltered.get(position);
        }

        @Override
        public long getItemId(int position) {
            return position;
        }

        @Override
        public View getView(int position, View convertView, ViewGroup parent) {
            if (convertView == null) {
                convertView = LayoutInflater.from(context).inflate(android.R.layout.simple_list_item_multiple_choice, parent, false);
                // We want custom layout with icon, but simple_list_item_multiple_choice is easiest for checkboxes.
                // Let's create a very simple one instead.
                convertView = LayoutInflater.from(context).inflate(R.layout.app_item, parent, false);
            }

            AppInfo info = appsFiltered.get(position);
            TextView text = convertView.findViewById(R.id.app_name);
            TextView pkg = convertView.findViewById(R.id.app_package);
            ImageView icon = convertView.findViewById(R.id.app_icon);
            CheckBox cb = convertView.findViewById(R.id.app_check);

            text.setText(info.name);
            pkg.setText(info.packageName);
            icon.setImageDrawable(info.icon);
            cb.setChecked(info.selected);

            convertView.setOnClickListener(v -> {
                info.selected = !info.selected;
                cb.setChecked(info.selected);
            });

            return convertView;
        }

        @Override
        public Filter getFilter() {
            if (filter == null) {
                filter = new AppFilter();
            }
            return filter;
        }

        private class AppFilter extends Filter {
            @Override
            protected FilterResults performFiltering(CharSequence constraint) {
                FilterResults results = new FilterResults();
                if (constraint == null || constraint.length() == 0) {
                    results.values = appsOriginal;
                    results.count = appsOriginal.size();
                } else {
                    String query = constraint.toString().toLowerCase();
                    List<AppInfo> filtered = new ArrayList<>();
                    for (AppInfo info : appsOriginal) {
                        if (info.name.toLowerCase().contains(query) || info.packageName.toLowerCase().contains(query)) {
                            filtered.add(info);
                        }
                    }
                    results.values = filtered;
                    results.count = filtered.size();
                }
                return results;
            }

            @Override
            @SuppressWarnings("unchecked")
            protected void publishResults(CharSequence constraint, FilterResults results) {
                appsFiltered = (List<AppInfo>) results.values;
                notifyDataSetChanged();
            }
        }
    }
}

File: app/src/main/java/net/typeblog/socks/AppSelector.kt
package net.typeblog.socks

import android.app.Activity
import android.app.AlertDialog
import android.content.Context
import android.content.pm.ApplicationInfo
import android.content.pm.PackageManager
import android.graphics.drawable.Drawable
import android.text.Editable
import android.text.TextWatcher
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.BaseAdapter
import android.widget.CheckBox
import android.widget.EditText
import android.widget.Filter
import android.widget.Filterable
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.ListView
import android.widget.TextView

class AppSelector {

    fun interface OnAppSelectedListener {
        fun onAppSelected(appList: String)
    }

    companion object {
        fun show(activity: Activity, currentApps: String?, listener: OnAppSelectedListener) {
            val pm = activity.packageManager
            val packages = pm.getInstalledApplications(PackageManager.GET_META_DATA)
            val appList = mutableListOf<AppInfo>()

            val selectedPackages = mutableSetOf<String>()
            if (!currentApps.isNullOrEmpty()) {
                for (p in currentApps.split("\n")) {
                    selectedPackages.add(p.trim())
                }
            }

            for (packageInfo in packages) {
                val info = AppInfo()
                info.name = packageInfo.loadLabel(pm).toString()
                info.packageName = packageInfo.packageName
                info.icon = packageInfo.loadIcon(pm)
                info.selected = selectedPackages.contains(info.packageName)
                appList.add(info)
            }

            appList.sortWith(Comparator { o1, o2 ->
                o1.name.compareTo(o2.name, ignoreCase = true)
            })

            val layout = LinearLayout(activity).apply {
                orientation = LinearLayout.VERTICAL
            }

            val searchBox = EditText(activity).apply {
                hint = activity.getString(R.string.search_apps)
                isSingleLine = true
            }
            layout.addView(searchBox)

            val listView = ListView(activity)
            val adapter = AppAdapter(activity, appList)
            listView.adapter = adapter

            val listParams = LinearLayout.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT, 0, 1.0f
            )
            layout.addView(listView, listParams)

            searchBox.addTextChangedListener(object : TextWatcher {
                override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {}
                override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {
                    adapter.filter.filter(s)
                }
                override fun afterTextChanged(s: Editable?) {}
            })

            AlertDialog.Builder(activity)
                .setTitle(R.string.adv_app_selector)
                .setView(layout)
                .setPositiveButton(android.R.string.ok) { _, _ ->
                    val result = appList.filter { it.selected }
                        .joinToString("\n") { it.packageName }
                    listener.onAppSelected(result)
                }
                .setNegativeButton(android.R.string.cancel, null)
                .show()
        }
    }

    private class AppInfo {
        var name: String = ""
        var packageName: String = ""
        lateinit var icon: Drawable
        var selected: Boolean = false
    }

    private class AppAdapter(
        private val context: Context,
        private val appsOriginal: List<AppInfo>
    ) : BaseAdapter(), Filterable {

        private var appsFiltered: List<AppInfo> = appsOriginal
        private var filter: AppFilter? = null

        override fun getCount(): Int = appsFiltered.size

        override fun getItem(position: Int): Any = appsFiltered[position]

        override fun getItemId(position: Int): Long = position.toLong()

        override fun getView(position: Int, convertView: View?, parent: ViewGroup?): View {
            val v = convertView ?: LayoutInflater.from(context)
                .inflate(R.layout.app_item, parent, false)

            val info = appsFiltered[position]
            val text = v.findViewById<TextView>(R.id.app_name)
            val pkg = v.findViewById<TextView>(R.id.app_package)
            val icon = v.findViewById<ImageView>(R.id.app_icon)
            val cb = v.findViewById<CheckBox>(R.id.app_check)

            text.text = info.name
            pkg.text = info.packageName
            icon.setImageDrawable(info.icon)
            cb.isChecked = info.selected

            v.setOnClickListener {
                info.selected = !info.selected
                cb.isChecked = info.selected
            }

            return v
        }

        override fun getFilter(): Filter {
            if (filter == null) {
                filter = AppFilter()
            }
            return filter!!
        }

        private inner class AppFilter : Filter() {
            @Suppress("UNCHECKED_CAST")
            override fun performFiltering(constraint: CharSequence?): FilterResults {
                val results = FilterResults()
                if (constraint.isNullOrEmpty()) {
                    results.values = appsOriginal
                    results.count = appsOriginal.size
                } else {
                    val query = constraint.toString().lowercase()
                    val filtered = appsOriginal.filter {
                        it.name.lowercase().contains(query) ||
                                it.packageName.lowercase().contains(query)
                    }
                    results.values = filtered
                    results.count = filtered.size
                }
                return results
            }

            @Suppress("UNCHECKED_CAST")
            override fun publishResults(constraint: CharSequence?, results: FilterResults) {
                appsFiltered = results.values as List<AppInfo>
                notifyDataSetChanged()
            }
        }
    }
}
